File: swift/Sources/Fizzy/TokenProvider.swift
/// A type that provides access tokens for API authentication.
///
/// Implement this protocol to supply tokens to the SDK.
/// Tokens are requested before each API call, allowing for
/// automatic token refresh.
public protocol TokenProvider: Sendable {
    /// Returns the current access token.
    ///
    /// Called before each API request. Implementations may cache
    /// the token and refresh it when expired.
    func accessToken() async throws -> String
}

/// A token provider that returns a static token string.
///
/// Suitable for scripts, testing, and scenarios where tokens
/// don't need to be refreshed.
///
/// ```swift
/// let client = FizzyClient(
///     accessToken: "your-token",
///     userAgent: "my-app/1.0 (you@example.com)"
/// )
/// ```
public struct StaticTokenProvider: TokenProvider, Sendable {
    private let token: String

    /// Creates a static token provider with the given token.
    public init(_ token: String) {
        self.token = token
    }

    public func accessToken() async throws -> String {
        token
    }
}

File: kotlin/sdk/src/commonMain/kotlin/com/basecamp/fizzy/TokenProvider.kt
package com.basecamp.fizzy

/**
 * Provides access tokens for authenticating with the Fizzy API.
 *
 * Implementations can provide static tokens or dynamic token refresh logic.
 *
 * ```kotlin
 * // Static token
 * val provider = StaticTokenProvider("your-token")
 *
 * // Dynamic token with refresh
 * val provider = TokenProvider { refreshAccessToken() }
 * ```
 */
fun interface TokenProvider {
    /** Returns the current access token. */
    suspend fun accessToken(): String
}

/** A [TokenProvider] that always returns the same token. */
class StaticTokenProvider(private val token: String) : TokenProvider {
    init {
        require(token.isNotBlank()) { "Access token must not be blank" }
    }

    override suspend fun accessToken(): String = token
}
